### ISSUE TITLE
Question: Smart constructor pattern / composable validations

### ISSUE THREAD

**Issue by jedesroches** [konform-kt/konform#96]
Hello, and thank you for this nice library. Coming from the FP world, I am trying to use it with the smart constructor pattern and, not having a bind (or flatMap, or andThen, or anything you call `(ValidationResult<T1> -> (T1 -> ValidationResult<T2>)) -> ValidationResult<T2>`), am having a bit of trouble: I'd be happy to open a PR with a bind implementation, but first I'm thinking maybe I'm using this wrong.

Using Konform, I've managed to build the following inline class, who'se type represents a proof of being valid (I don't want to have `Valid<x>`'s all over my domain code), which works well enough:

```kotlin
@JvmInline
value class CustomerID private constructor(val value: String) {
    companion object {
        operator fun invoke(input: String): ValidationResult<CustomerID> = Validation {
            pattern("^[A-Z0-9]{32}$") hint "Customer ID <value> does not match `^[A-Z0-9]{32}$`"
        }(CustomerID(input))
    }
}
```

But then when I have a class that is using the above value class as a member, I cannot compose both validations. Given:

```kotlin
data class CustomerDto(
    val id: String,
    val name: String,
    val whatever: String,
)

data class Customer(
    val id: CustomerID,
    val name: CustomerName,
    val whatever: CustomerWhatever,
)
```

Going from the first above to the second, I cannot do something like `CustomerID(id).andThen { id -> ... }` to build the final object. But I have to validate the `CustomerID` before passing it to the `Customer` constructor (since the `CustomerID` type enforces validity of the data it contains), and so I am left with no choice but to validate by hand the different elements, and compose the errors manually.

_What is the suggested usage pattern for Konform to have both type-level guarantees of data validity when passing values around and composability of validations to build more complex objects ?_

I hope I've managed to explain my question correctly, and I'm looking forward to your answer! Have a great day :smiley:

**Comment by dhoepelman** [comment]
I'm not 100% sure I follow, but I think it would be good enough for `ValidationResult` to be a functor, i.e. implement a proper `map`. Then you can do:

```kotlin
val customer: Customer = ...
val dto: ValidationResult<CustomerDto>= customerValidation(customer).map { it.toDto() }
```

The current (0.4.0) `map` doesn't properly work, fixed in 0.5.0. I am currently trying to release it (just took over maintaining, ironing out the release process)

To help you: In the JVM world we usually call haskell's fmap `map`, and monadic bind `flatMap`. Monadic unit doesn't have a standard convention, sometimes it's `just`. Trying to find these in types and using this nomenclature for issues will make things more clear. You will probably also enjoy [Arrow](https://arrow-kt.io/)

**Comment by dhoepelman** [comment]
Closing this as it's not an issue per se, feel free to reply or open a new issue with a more directed question/problem
